README

Running the Test application
----------------------------

The test scripts run a minimal app called 'blah',
against a sample SQLite database.

To actually run the test app, do this:

EXAMPLEDB=t/example.db t/blah/script/blah daemon

Building a distribution
-----------------------
perl Makefile.PL
make test
make manifest
make dist
mojo cpanify -u USER -p PASS *Tables*tar.gz



CHANGES
commit 28e3529e25d7bdc4c926dcd988696881e396b015
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Apr 4 12:27:26 2021 +0100

    Build version 1.07

commit 523b62a6f21c177c092321d4316b51909d1e57a0
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Apr 4 12:22:55 2021 +0100

    fix build instructions

commit b4694899648b6208d641b0048d0ec2ffd252a236
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Apr 4 12:12:50 2021 +0100

    Built version 1.06

commit 808a137f0fcbdf025ccdb7e9c65f6b3c969bee97
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Apr 4 12:09:47 2021 +0100

    support new Mojolicious backward-breaking "format" rules

commit 8b43a4bc86a0fde297588636a5b01d1af03312eb
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sat May 16 12:53:25 2020 +0100

    Built version 1.05

commit 9b1b03c97bf3282ca02650dac922576d479bd590
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sat May 16 12:52:14 2020 +0100

    Build version 1.05

commit 623482048a8a3c97581007128a3d6813534eec54
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Thu May 14 17:16:35 2020 +0100

    first childset display should be ordered

commit 36c543f85f62d55dfc2112ff8aacdcba497e6676
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Wed May 13 10:49:49 2020 +0100

    make logourl conditional

commit abe4f6ef46e3c2732671701f66628c619fc9838f
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Fri Feb 28 17:43:47 2020 +0000

    readme

commit a97effb3e8aedec57be68135a1c2fea3cdb0bb39
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Fri Feb 28 17:39:00 2020 +0000

    Built version 1.04

commit d64d4ed106793f854e29b76c2e77844febbdd73e
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Fri Feb 28 17:34:52 2020 +0000

    Build version 1.04

commit 9c3bbbb78dc2c754ea6fb50035d7001c3682095b
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Fri Feb 28 17:28:24 2020 +0000

    handling deploy-to-root

commit 8b1a717d556763ba0290c5425c44de543ca4d572
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sat Feb 22 17:05:47 2020 +0000

    comment only.. more reliable test for forwarded

commit 231aec3171bf5d763e1cb1a3834ba825095d8772
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Thu Dec 19 18:34:01 2019 +0000

    improve distribution instructions

commit 7e4ebd976a1ddc89b8edb07c07fad0d07940f7c4
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Thu Dec 19 18:30:29 2019 +0000

    Build version 1.03

commit b4763d4458b9ac74c3f4c7d8ef8fac4dc08d65cb
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Thu Dec 19 18:26:31 2019 +0000

    setup shipped->base earlier

commit 590ae3059e837f919698778ff7897866248674e8
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 8 23:30:05 2019 +0000

    tweak urls for exception templates

commit 4049bf4e6c301c81b1b83293b809a0c276f49d37
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 8 20:33:15 2019 +0000

    travis: specify Perl 5.28

commit fb7b0635fb4e33f985e78135a8970ea20ffa6835
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 8 20:21:41 2019 +0000

    update CHANGES

commit 5fb4c31ec4513a18094de8bea8d1b99ad39f008b
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 8 20:20:37 2019 +0000

    remove noise files

commit 029341323bf21d69202435e43088b039a338ef37
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 8 20:13:10 2019 +0000

    jscript patches for base handling

commit 69f89a57f54c8353d9e467a6fd50e39753bf1df0
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Thu Apr 18 20:23:22 2019 +0100

    Build version 1.01

commit ec768a7ffa83bf148f20e7f0cf4ce9987ac3f0b3
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Thu Apr 18 20:18:03 2019 +0100

    make a Tables app deployable under a path

commit 189c85821063222960712db85e063558fe5fee77
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Fri Jul 7 14:51:04 2017 +0000

    Build version 0.08

commit 1a03a303bfbe4ddd4c948bfe1c2c7dcfba63027c
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Fri Jul 7 14:48:08 2017 +0000

    build child-link using compound-ids

commit a7d8801b18b51a881a60196e5f0aa6698acba0f5
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Mon Dec 19 19:21:26 2016 +0000

    Build version 0.07 with dist

commit ec29ef0f84ae2e0882a5be34f281a103b7f12f1d
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Mon Dec 19 19:20:32 2016 +0000

    Build version 0.07

commit 996a961dea152def4f27f77acf303516bba1e53e
Merge: 003a27f 3606a70
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Mon Dec 19 19:11:52 2016 +0000

    Merge remote-tracking branch 'origin/master'

commit 003a27f80b4ee226d7bd359cccd164ae33657759
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Mon Dec 19 19:10:28 2016 +0000

    Build version 0.06

commit 88b7169ad43608d87cf02b06055896f8bf477837
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Mon Dec 19 19:09:09 2016 +0000

    tweak build instructions

commit 3606a7032ecf4335d709499e98207be1afb16b22
Merge: 39fcc81 821b4e6
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Mon Dec 19 19:01:01 2016 +0000

    Merge pull request #2 from manwar/fix-missing-prereqs
    
    Fix missing prereqs as reported by CPANTS.

commit 821b4e68a07f48d36be921453e2039f71ac03af9
Author: Mohammad S Anwar <mohammad.anwar@yahoo.com>
Date:   Mon Dec 19 11:45:17 2016 +0000

    Fix missing prereqs as reported by CPANTS.
    http://cpants.cpanauthors.org/dist/Mojolicious-Plugin-Tables

commit 39fcc812f03fb1ca4d2aca4c16bcdb537db73e2d
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 18 23:34:42 2016 +0000

    Build version 0.05

commit 4e4c6db5a0e6a15967ceac0765ae999fbbabc7e7
Author: Frank Carnovale <frank.carnovale@gmail.com>
Date:   Sun Dec 18 23:22:58 2016 +0000


t/blah/lib/Blah.pm
package Blah;
use Mojo::Base 'Mojolicious';
use StuffDB;

sub startup {
  my $self = shift;

  $self->plugin(Tables => { connect_info=>["dbi:SQLite:$ENV{EXAMPLEDB}", '', ''] });

}

1;


t/blah/lib/StuffDB.pm
package StuffDB;

use strict;
use warnings;

use base qw/Mojolicious::Plugin::Tables::Model/;

sub connect_info { [ 'dbi:Pg:dbname="stuff"', '', '' ] }

sub glossary { +{
    abs       => 'Absolute',
    avg       => 'Average',
    cap       => 'Capitalization',
    cd        => 'Code',
    dyngrp    => 'Dynamic Group',
    dyngrps   => 'Dynamic Groups',
    id        => 'Identifier',
    ip        => 'IP Number',
    isin      => 'ISIN',
    fca       => 'FCA',
    lse       => 'LSE',
    mkt       => 'Market',
    pargrp    => 'Parent Group',
    pct       => '%',
    prc       => 'Price',
    subgroups => 'Sub-Groups',
    feedbacks => 'Feedback',
    tidm      => 'TIDM',
    ts        => 'Timestamp',
    var       => 'Variance',
    vol       => 'Volume',
    marketsegmentcode => 'Market Segment Code',
    marketsectorcode  => 'Market Sector Code',
    subsector         => 'Sub-Sector',
} }

sub input_attrs { +{
    var_pct => { min=>-100, max=>100 },
    vol_pct => { min=>-100, max=>500 },
    details => { cols=>60 },
    name    => { size=>80 },
    aud     => { size=>80 },
    azp     => { size=>80 },
    description => { size=>80 },
    hd      => { size=>80 },
    picture => { size=>80, type=>'url' },
    email   => { size=>40, type=>'email' },
    email_verified => { type=>'checkbox' },
    ts      => { step=>1 },
    joined  => { step=>1 },
} }

sub rel_name_map { +{
    AssetRange => { range_age => 'range' },
    Asset      => { symbol    => 'lse_security' },
    Dyngrp     => { dyngrps   => 'subgroups' },
    Trader     => { authority_granters => 'grants_to',
                    authority_traders  => 'audit_trail' },
} }

1;



t/blah/lib/StuffDB/Asset.pm
package StuffDB::Asset;

use strict;
use warnings;

sub stringify {
    my $self = shift;
    my $latest_str = '';
    if (my $latest = $self->latest) {
        $latest_str = " $latest";
        if (my $var_pct = $self->var_pct) {
            $latest_str .= sprintf ' (%+.2f%%)', $var_pct
        }
    }
    if (my $secty = $self->lse_security) {
        return sprintf '%s %s (%s)%s', $secty->company_name, $secty->security_name, $self->symbol, $latest_str
    }
    return sprintf '%s (%s)%s', $self->name, $self->dataset_code, $latest_str
}

sub short {
    my $self = shift;
    sprintf '%s (%s)', $self->name, $self->dataset_code
}

sub asset_ranges_descending {
    my $self = shift;
    my $cond = shift;
    $self->asset_ranges->search($cond, {order_by=>{-desc=>'range_age'}});
}

sub fn { # filename
    my $self = shift;
    sprintf "$ENV{QUANDL}/%s/%s.out", $self->database->id, $self->dataset_code;
}

1;



Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojolicious::Plugin::Tables',
  VERSION_FROM => 'lib/Mojolicious/Plugin/Tables.pm',
  ABSTRACT     => 'Quickstart and grow a Tables-Maintenance Webapp',
  AUTHOR       => 'Frank Carnovale <frankc@cpan.org>',
  LICENSE      => 'artistic_2',
  PREREQ_PM    => {'Mojolicious' => '6.44',
                   'DBIx::Class::Schema::Loader' => '0.07045',
                   'DBIx::Class::Schema::Loader::Dynamic' => '1.05' },
  EXE_FILES    => ['bin/tables'],
  test         => {TESTS => 't/*.t'},
  META_MERGE   => {
    dynamic_config => 0,
    'meta-spec'    => {version => 2},
    no_index       => {directory => ['t']},
    prereqs        => {runtime => {requires => {perl => '5.010001'}}},
    resources    => {
        bugtracker => {web => 'https://github.com/frank-carnovale/Mojolicious-Plugin-Tables/issues'},
        license    => ['http://www.opensource.org/licenses/artistic-license-2.0'],
        repository => {
            type => 'git',
            url  => 'https://github.com/frank-carnovale/Mojolicious-Plugin-Tables.git',
            web  => 'https://github.com/frank-carnovale/Mojolicious-Plugin-Tables',

        },
    },
  },
);


lib/Mojolicious/Plugin/Tables.pm
package Mojolicious::Plugin::Tables;
use Mojo::Base 'Mojolicious::Plugin';

use File::Basename 'dirname';
use File::Spec::Functions 'catdir';

use Mojolicious::Plugin::Tables::Model;

our $VERSION = '1.07';

sub register {
    my ($self, $app, $conf) = @_;

    my $log = $app->log;

    $app->helper(add_stash => sub {
        my ($c, $slot, $val) = @_;
        push @{$c->stash->{$slot} ||= []}, $val;
    });

    $app->helper(add_flash => sub {
        my ($c, $slot, $val) = @_;
        push @{$c->session->{new_flash}{$slot} ||= []}, $val;
    });

    $app->helper(shipped => sub {
        # special stash slot for sending json-able structs to js;
        # get/set logic here cloned from Mojo::Util::_stash.
        my $c = shift;
        my $shipped = $c->stash->{shipped} ||= {};
        return $shipped unless @_;
        return $shipped->{$_[0]} unless @_>1 || ref $_[0];
        my $values = ref $_[0]? $_[0] : {@_};
        @$shipped{keys %$values} = values %$values;
    });

    $app->config(default_theme=>'redmond');
    $app->defaults(layout => $conf->{layout} || 'tables');

    my $model_class = $conf->{model_class} ||= 'Mojolicious::Plugin::Tables::Model';
    eval "require $model_class" or die;

    $model_class->log($log);
    my $schema = $model_class->setup($conf);
    my $model = $schema->model;
    $app->config(model=>$model);

    $app->hook(before_dispatch => sub {
        my $c = shift;
        # Move first part and slash from path to base path when deployed under a path
        if ($c->req->headers->to_string =~ /X-Forwarded/) {
            my $part0 = shift @{$c->req->url->path->leading_slash(0)};
            push @{$c->req->url->base->path->trailing_slash(0)}, $part0;
            $c->shipped(urlbase => $c->req->url->base);
        } else {
            $c->shipped(urlbase => '');
        }
        # capture https into base
        if ($c->req->headers->header('X-Forwarded-HTTPS')
        || ($c->req->headers->header('X-Forwarded-Proto')||'') eq 'https') {
            $c->req->url->base->scheme('https')
        }
    });
 
    my $plugin_resources = catdir dirname(__FILE__), 'Tables', 'resources';
    push @{$app->routes->namespaces}, 'Mojolicious::Plugin::Tables::Controller';
    push @{$app->renderer->paths}, catdir($plugin_resources, 'templates');
    push @{$app->static->paths},   catdir($plugin_resources, 'public');

    # Arrange for custom table-specific template overrides, when present.
    # We do this by wrapping the standard ep-handler with a
    # pre-processor to test for existence of the custom template.
    # Any internal caching by EP renderer is preserved even for overrides.

    my $handlers = $app->renderer->handlers;
    my $base_ep  = $handlers->{ep};
    $handlers->{ep} = sub {
        my ($renderer, $c, $output, $options) = @_;
        {
            my $table           = $c->stash('table') || last;
            my $custom_template = "$table/$options->{template}";
            my $custom_path     = $renderer->template_path ({%$options,
                                             template=>$custom_template}) || last;
            $options->{template} = $custom_template;
            $log->debug("custom 'tables' template at $custom_path") if $ENV{TABLES_DEBUG};
        }
        $base_ep->(@_)
    };

    my $r = $app->routes;
    $r->get('/' => sub{shift->redirect_to('tables')}) unless $conf->{nohome};

    my @crud_gets  = (qw/view edit del nuke navigate/);
    my @crud_posts = (qw/save/);
    my $fmts       = [format=>[qw/html json/]];

    for ($r->under('tables')                          ->to('auth#ok'   )) {
        for ($_->under()                              ->to('tables#ok' )) {
            $_->get                                   ->to('#page'     );
            for ($_->under(':table')                  ->to('#table_ok' )) {
                $_->any('/'=>$fmts)                   ->to('#table', format=>'html' );
                $_->get('add')                        ->to('#add'      );
                $_->post('save')                      ->to('#save'     );
                for ($_->under(':id')                 ->to('#id_ok'    )) {
                    my $r = $_;
                    $r->get( $_=>$fmts)               ->to("#$_", format=>'html') for @crud_gets;
                    $r->post($_=>$fmts)               ->to("#$_", format=>'html') for @crud_posts;
                    $r->get('add_child/:child'=>$fmts)->to('#view', format=>'html'     );
                    $r->any(':children'=>$fmts)       ->to('#children', format=>'html' );
                }
            }
        }
    }

    my @tablist = @{$model->{tablist}};
    $log->info ("'Tables' Framework enabled.");
    $log->debug("Route namespaces are..");
    $log->debug("--> $_") for @{$app->routes->namespaces};
    $log->debug("Renderer paths are..");
    $log->debug("--> $_") for @{$app->renderer->paths};
    $log->debug("Static paths are..");
    $log->debug("--> $_") for @{$app->static->paths};
    $log->info ("/tables routes are in place. ".scalar(@tablist)." tables available");
    $log->debug("--> $_") for @tablist;

    $log->debug($app->dumper($model)) if $ENV{TABLES_DEBUG};

}

1;
__END__

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Tables -- Quickstart and grow a Tables-Maintenance application

=head1 SYNOPSIS

    # in a new Mojolicious app..
 
    $app->plugin( Tables => {connect_info=>['dbi:Pg:dbname="mydatabase"', 'uid', 'pwd']} );

    # when ready to grow..

    $app->plugin( Tables => {model_class => 'StuffDB'} );

=head1 DESCRIPTION

L<Mojolicious::Plugin::Tables> is a L<Mojolicious> plugin which helps you grow a high-featured web app by
starting with basic table / relationship maintenance and growing by overriding default behaviours.

By supplying 'connect_info' (DBI-standard connect parameters) you get a presentable database maintenance
web app, featuring full server-side support for paged lists of large tables using
'Datatables' (datatables.net), plus ajax handling for browsing 1-many relationships, plus
JQueryUI-styled select lists for editing many-to-1 picklists.

By supplying your own Model Class you can override most of the default behaviours
and build an enterprise-ready rdbms-based web app.

By supplying your own templates for context-specific calls you can start to give your app a truly 
specialised look and feel.

=head1 STATUS

This is an early release.  Guides and more override-hooks coming Real Soon Now.

=head1 GROWTH PATH

=head2 Ground Zero Startup

    tables dbi-connect-params..

The 'tables' script is supplied with this distribution.  Give it standard DBI-compatible parameters on the 
commandline and it will run a minimal web-server at localhost:3000 to allow maintenance on the named database.

=head2 Day One

In your Mojolicious 'startup'..

    $self->plugin(Tables => { connect_info => [ connect-param, connect-param, .. ] };

Add this line to a new Mojolicious app then run it using any Mojo-based server (morbo, prefork, hypnotoad..) to achieve exactly the same functionality as the 'tables' script.

=head2 Day Two

    # templates/:table/{view,edit,dml,page}.html.ep
    # e.g. 
    # templates/artist/view.html.ep
    <h1>Artist: <%= $row %></h1>

For any :table in your database, create override-templates as required.
e.g. The above code will give a very different look-and-feel when viewing a single Artist, but all other
pages are unchanged.  For better examples and to see which stash-variables are available, start by 
copying the distribution templates from ../Plugin/Tables/resources/templates into your own template area.

=head2 Infinity and Beyond

    $self->plugin(Tables => { model_class => 'MyDB' } );

Prepare your own model_class to override the default database settings which "Tables" has determined from the 
database.  This class (and its per-table descendants) can be within or without your Mojolicious application.
C<model_class> implements the specification given in L<Mojolicious::Plugin::Tables::Model>.
This lets you start to customise and grow your web app.

=head1 CONFIGURATION

    $app->plugin(Tables => $conf) 

Where the full list of configuration options is:

=head3 layout

Provide an alternative 'wrapper' layout; typically the one from your own application.  If you prepare one of these you will need
to include most of the layout arrangements in the Day-One layout, i.e. the one at resources/templates/layouts/tables.html.ep.
The easiest approach is to start by copying the packaged version into your own application and then change its look and feel to
suit your requirements.

=head3 nohome

Unless you supply a true value for this, an automatic redirection will be in place from the root of your app to the '/tables' path.
The redirection is there to make day-one functionality easy to find, but once your app grows you will not want this redirection.

=head3 model_class

See 'Customising Model Class'

=head3 connect_info

See 'Day One'

=head3 default_theme

To experiment with completely different colour themes, choose any standard JQueryUI theme name or "roll" your own as described here 
L<http://jqueryui.com/themeroller/>.   Our default is 'Redmond'.

=head1 DEBUGGING

To generate detailed trace info into the server log, export TABLES_DEBUG=1.

=head1 CAVEAT

We use dynamically-generated DBIx::Class classes.  This technique does not scale well for very large numbers
of tables.  Previous (private) incarnations of this Framework used specially prepared high-performance versions of 
Class::DBI::Loader to speed this up.  So that speeding-up at start-time is a TODO for this DBIx::Class-based release.

=head1 SOURCE and ISSUES REPOSITORY

Open-Sourced at Github: L<https://github.com/frank-carnovale/Mojolicious-Plugin-Tables>.  Please use the Issues register there.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016, Frank Carnovale <frankc@cpan.org>

This program is free software, you can redistribute it and/or modify it under the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Mojolicious>, L<DBIx::Class::Schema::Loader>, L<Mojolicious::Plugin::Tables::Model>

=cut


lib/Mojolicious/Plugin/Tables/Model/Row.pm
package Mojolicious::Plugin::Tables::Model::Row;

use strict;
use warnings;

use overload
    '""'     => sub { shift->stringify_safely },
    'bool'   => sub { 1 },
    fallback => 1;

# errors during stringification of a db object can sometimes trigger 
# deep recursion by well-meaning error messages deep in ORM internals.

sub stringify_safely {
    my $self = shift;
    return eval { $self->stringify } // do {
        my $err = $@;
        my $class  = ref $self;
        my $fallback = "[*$class]".$self->id;
        $self->log->error("stringifying $fallback: $err");
        $fallback
    }
}

# override this for each ResultClass.

sub stringify {
    my $self  = shift;
    my $cd = $self->can('cd')
                    ? (':'.$self->cd)
                    : $self->compound_ids;
    my $ds = $self->can('description')? $self->description:
             $self->can('name'       )? $self->name:
             '';
    my $label = $self->result_source->source_name;
    sprintf '[%s] %s%s', $label, $cd, ($ds? " - $ds":'')
}

sub log { shift->result_source->schema->log }

sub compound_ids {
    my $self = shift;
    join ('-|-', map { $self->get_column($_) } $self->primary_columns);
}

sub present {
    my ($self, $column, $info, %opts) = @_;
    my $class = ref $self;
    #$self->log->debug("present $column for $class using " . Dumper($info));
    my $type = $info->{data_type} || 'varchar';

    my $val = $self->$column // return;

    for ($type) {
        /timestamp/ && return ($opts{foredit}? $val->iso8601: $val->strftime('%F %T'));
        /date/      && return ($val->ymd);
        /boolean/   && return ($val?'Yes':'No');
    }
    return $val
}

# generate the full pick-list that lets the fk $column pick from its parent,
# in a structure suitable for the 'select_field' tag.  This version gets all
# (limited by safety check) but inherited versions are expected to do more
# context-sensitive filtering.  Will work as a class method.
sub options {
   #my ($self, $column, $cinfo, $pinfo, $schema, $bytable) = @_;
    my ($xxx1, $xxxxx2, $xxxx3, $pinfo, $schema, $bytable) = @_;
    my $ptable   = $pinfo->{ptable};
    my $ptabinfo = $bytable->{$ptable};
    my $psource  = $ptabinfo->{source};
    my $prs      = $schema->resultset($psource);
    my $where    = {};
    my $attrs    = {rows=>200};
    my @options  = map { 
                        [ "$_" => $_->id ]
                    } $prs->search($where, $attrs);
    return \@options;
}

sub nuke {
    my $self   = shift;
    my $s      = $self->result_source;
    my $schema = $s->schema;

    my @collections = grep { $s->relationship_info($_)->{attrs}{accessor} eq 'multi' }
                           $s->relationships;

    my $i = 0;
    $schema->txn_do( sub {
        for my $collection (@collections) {
            $i += $_->nuke for $self->$collection->all;
        }
        $self->delete;
        ++$i
    } )
}

1;



lib/Mojolicious/Plugin/Tables/Model.pm
package Mojolicious::Plugin::Tables::Model;

use strict;
use warnings;

use base qw/DBIx::Class::Schema/;
use DBIx::Class::Schema::Loader::Dynamic;

__PACKAGE__->mk_group_accessors(inherited => qw/log model/);

sub setup {
    my ($class, $conf) = @_;
    if (my $connect_info = $conf->{connect_info}) {
        $class->connect_info($connect_info)
    } else {
        die "Provide connect_info either as a config value or an override"
            unless $class->connect_info
    }

    # to return a schema object-ref here say 'connect' instead of 'connection'.
    my $schema = $class->connection(@{$class->connect_info});

    DBIx::Class::Schema::Loader::Dynamic->new(
        left_base_classes   => $class->row_base,
        rel_name_map        => $class->rel_name_map,
        custom_column_info  => sub { $class->custom_column_info(@_) },
        naming              => 'v8',
        use_namespaces      => 0,
        schema              => $schema,
        %{$conf->{loader_opts}||{}},
    )->load;

    $schema->model($schema->_model);

    return $schema;
}

sub row_base { 'Mojolicious::Plugin::Tables::Model::Row' }

sub glossary { +{ id => 'Identifier' } }

sub input_attrs { +{ name => { size=>80 } } }

sub make_label {
    my $self = shift;
    my $name = shift;
    my @label = split '_', $name;
    for (@label) {
        $_ = $self->glossary->{$_}, next if $self->glossary->{$_};
        $_ = ucfirst
    }
    join(' ', @label)
}

sub custom_column_info {
    my ($class, $table, $column, $column_info) = @_;
    my $info = { label => $class->make_label($column) };
    my $attrs1;
    for ($column_info->{data_type}) {
        $attrs1 =
            /numeric|integer/ ? {type=>'number'} :
            /timestamp/       ? {type=>'datetime-local'} :
            /date|time/       ? {type=>$_} :
            {};
    }
    my $attrs2 = $class->input_attrs->{$column} || {};
    $info->{input_attrs} = {%$attrs1, %$attrs2} if keys(%$attrs1) || keys(%$attrs2);
    $info
};

sub rel_name_map { +{} }

sub _model {
    my $schema  = shift;

    my @tablist = ();
    my %bytable = ();
    #my $log     = $schema->log;
    #$log->debug("$schema is building its model");
    for my $source (sort $schema->sources) {
        my $s = $schema->source($source);
        my @has_a;
        my %has_many;
        for my $rel ($s->relationships) {
            my $info   = $s->relationship_info($rel);
            my $ftable = $info->{class}->table;
            my $attrs  = $info->{attrs};
            my $card   = $attrs->{accessor};
            if ($card eq 'single') {
                my $fks = $attrs->{fk_columns};
                my @fks = keys %$fks;
                push @has_a, { fkey=>$fks[0], parent=>$rel, label=>$schema->make_label($rel), ptable=>$ftable }
                    if @fks == 1
            } elsif ($card eq 'filter') {
                my @ffkeys = keys %{$info->{cond}};
                if (@ffkeys == 1) {
                    (my $cfkey = $ffkeys[0]) =~ s/^foreign\.//;
                    push @has_a, { fkey=>$cfkey, parent=>$rel, label=>$schema->make_label($rel), ptable=>$ftable }
                } else {
                    warn __PACKAGE__." model: $source: $rel: multi-barrelled M-1 keys not supported\n"
                }
            } elsif ($card eq 'multi') {
                my $fsource_name = $info->{source};
                my $fsource      = $schema->source($fsource_name);
                my $fpkey        = join(',', $fsource->primary_columns);
                my @ffkeys       = keys %{$info->{cond}};
                if (@ffkeys == 1) {
                    (my $cfkey = $ffkeys[0]) =~ s/^foreign\.//;
                    $has_many{$rel} = {ctable=>$ftable, cpkey=>$fpkey, cfkey=>$cfkey, label=>$schema->make_label($rel)};
                } else {
                    warn __PACKAGE__." model: $source: $rel: multi-barrelled 1-M keys not supported\n"
                }
            } else {
                warn __PACKAGE__." model: $source: $rel: strange cardinality: $card\n";
            }
        }
        my %bycolumn = map {
                          my %info = %{$s->column_info($_)};
                          delete $info{name};
                          /^_/ && delete $info{$_} for keys %info;
                          ( $_ => \%info )
                      } $s->columns;
        for (@has_a) {
            my $fkey = $_->{fkey};
            my $parent = delete $_->{parent};
            $bycolumn{$fkey}->{parent} = $parent if $bycolumn{$fkey};
            $bycolumn{$parent} = $_; # gets {fkey=>, label=>, ptable=>,}
        }
        my $pkeys   = [$s->primary_columns];
        my $pknum   = 0;
        for (@$pkeys) {
            $bycolumn{$_}{is_primary_key} = ++$pknum
        }
        my @columns = map { $_, $bycolumn{$_}{parent}? ($bycolumn{$_}{parent}): () }
                      $s->columns;
        my $label   = $schema->make_label($s->name);
        my $tabinfo = {
                source   => $source,
                columns  => \@columns,
                bycolumn => \%bycolumn,
                has_many => \%has_many,
                label    => $label,
                pkeys    => $pkeys,
        };
        push @tablist, $s->name;
        $bytable{$s->name} = $tabinfo;
    }
    return {schema=>$schema, tablist=>\@tablist, bytable=>\%bytable};
}

1;
__END__

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Tables::Model -- Customize the default generated Tables Model

=head1 SYNOPSIS

    $app->plugin( Tables => {model_class => 'StuffDB'} );

=head1 DESCRIPTION

By supplying your own Model Class you can override most of the default behaviours
and build an enterprise-ready rdbms-based web app.

Your model_class must inherit from Mojolicious::Plugin::Tables::Model.  So, e.g.

    package StuffDB;

    use strict;
    use warnings;

    use base qw/Mojolicious::Plugin::Tables::Model/;

    ...

=head2 model_class METHODS

Your model_class optionally implements any of the following methods to override the default behaviour.

=head3 connect_info

A sub returning an arrayref supplying your locally declared DBI parameters. e.g.

    sub connect_info { [ 'dbi:Pg:dbname="stuff"', '', '' ] }

which works for a Postgres database called 'stuff' accepting IDENT credentials.

=head3 glossary

A sub returning a hashref which maps "syllables" to displaynames.  "Syllables" are all the abbreviated words that
are separated by underscores in your database table and column names.  Any syllable by default is made into a nice
word by simply init-capping it, so the glossary only needs to supply non-obvious displaynames.
The mapping 'id=>Identifier' is built-in.
So for example with table or column names such as "stock", "active_stock", "stock_id", "stock_name",
"dyngrp_id", "mkt_ldr", "ccld_ts" we would supply just this

    sub glossary { +{
        ccld   => 'Cancelled',
        dyngrp => 'Dynamic Group',
        mkt    => 'Market',
        ldr    => 'Leader',
        ts     => 'Timestamp',
    } }

.. and we will see labels "Stock", "Active Stock", "Stock Identifier", "Stock Name",
"Dynamic Group Identifier", "Market Leader", "Cancelled Timestamp" in the generated application.
 
=head3 input_attrs

A sub returning a hashref giving appropriate html5 form-input tag attributes for any fieldname.  By default these
attributes are derived depending on field type and database length. But these can be overriden here, e.g.

    sub input_attrs { +{
        var_pct => { min=>-100, max=>100 },
        vol_pct => { min=>-100, max=>500 },
        name    => { size=>80 },
        picture => { size=>80, type=>'url' },
        email   => { size=>40, type=>'email' },
        email_verified => { type=>'checkbox' },
        ts      => { step=>1 },
    } }

=head3 rel_name_map

A sub returning a hashref which maps default generated relationship names to more appropriate choices.  More detail
in L<DBIx::Class::Schema::Loader>.  e.g.

    sub rel_name_map { +{
        AssetRange => { range_age => 'range' },
        Asset      => { symbol    => 'lse_security' },
        Dyngrp     => { dyngrps   => 'subgroups' },
        Trader     => { authority_granters => 'grants_to',
                        authority_traders  => 'audit_trail' },
    } }

=head2 ResultClass Methods

After creating a model_class as described above you will automatically be able to introduce 'ResultClass' classes
for any of the tables in your database.  Place these directly 'under' your model_class, e.g. if StuffDB is
your model_class and you want to introduce a nice stringify rule for the table 'asset', then you can
create the class StuffDB::Asset and give it just the stringify method. e.g. from http://octalfutures.com :

    package StuffDB::Asset;

    use strict;
    use warnings;

    sub stringify {
        my $self = shift;
        my $latest_str = '';
        if (my $latest = $self->latest) {
            $latest_str = " $latest";
            if (my $var_pct = $self->var_pct) {
                $latest_str .= sprintf ' (%+.2f%%)', $var_pct
            }
        }
        return sprintf '%s (%s)%s', $self->name, $self->dataset_code, $latest_str
    }

    1;

Any of these ResultClasses will inherit all methods of DBIx::Class::Row.
In addition these methods inherit all methods of Mojolicious::Plugin::Tables::Model::Row, namely:

=head3 stringify

It's recommended to implement this.  The stringification logic for this table.  The default implementation 
tries to use any columns such as 'cd' or 'description', and falls back to joining the primary keys.

=head3 present

Generate a presentation of a row->column value, formatted depending on type.  Args: column_name, a hash-ref containing schema info about that column, and a hashref containing context info (currently just 'foredit=>1').

=head3 options

Given: column_name, a hash-ref containing schema info about that column, a hash-ref containing info about the parent table, the full DBIX::Class schema, and a hash-ref containing schema information about all tables..

Generate the full pick-list that lets the fk $column pick from its parent,
in a structure suitable for the Mojolicious 'select_field' tag.  The default version simply lists all choices
(limited by safety row-count of 200) but inherited versions are expected to do more
context-sensitive filtering.  Works as a class method to support the 'add' context.

=head3 nuke

Perform full depth-wise destruction of a database record.  The default implementation runs an alghorithm to delete
all child records and finally delete $self.  Override this to prohibit (by dieing) or perform additional work.

=head3 all the rest

Of course, all the methods described at L<DBIx::Class::Row> can be overriden here.

=head1 SEE ALSO

L<Mojolicious::Plugin::Tables>

=cut

